src/dune_pkg/outdated_packages.ml

```diff
@@ -0,0 +1,226 @@
+      |> OpamFormula.atoms
+      |> List.exists ~f:(fun (name', _) -> OpamPackage.Name.equal pkg_name name'))
+  in
+  match Opam_repo.load_all_versions repos pkg_name with
+  | Error `Package_not_found -> Package_not_found pkg.info.name
+  | Ok versions ->
+    (match
+       List.max versions ~f:(fun x y ->
+         Ordering.of_int
+           (OpamPackage.Version.compare
+              (OpamFile.OPAM.version x)
+              (OpamFile.OPAM.version y)))
+     with
+     | Some newest_opam_file ->
+       let version = OpamFile.OPAM.version newest_opam_file in
+       (match
+          OpamPackage.Version.of_string pkg.info.version
+          |> OpamPackage.Version.compare version
+          |> Ordering.of_int
+        with
+        | Lt | Eq -> Package_is_best_candidate
+        | Gt ->
+          Better_candidate
+            { is_immediate_dep_of_local_package
+            ; name = pkg.info.name
+            ; newer_version = version |> OpamPackage.Version.to_string
+            ; outdated_version = pkg.info.version
+            })
+     | None -> Package_not_found pkg.info.name)
+;;
+
+let pp ~transitive ~lock_dir_path results =
+  explain_results_to_user ~transitive ~lock_dir_path results
+  @ List.concat_map
+      (outdated_packages results)
+      ~f:
+        (fun
+          { is_immediate_dep_of_local_package; name; outdated_version; newer_version } ->
+        (* If --transitive is passed, then we always print the available package. If
+           not, then we only print it if it is an immediate dependency of a local
+           package. *)
+        if transitive || is_immediate_dep_of_local_package
+        then
+          (* CR-someday alizter: Create table printing helpers in Console and use
+             those to align output. *)
+          [ Pp.hbox
+            @@ Pp.concat
+                 [ Pp.verbatim (Dune_lang.Package_name.to_string name)
+                 ; Pp.space
+                 ; Pp.tag
+                     (User_message.Style.Ansi_styles [ `Fg_bright_red ])
+                     (Pp.verbatim outdated_version)
+                 ; Pp.text " < "
+                 ; Pp.tag
+                     (User_message.Style.Ansi_styles [ `Fg_bright_green ])
+                     (Pp.verbatim newer_version)
+                 ]
+          ]
+        else [])
+;;
+
+let raise_errors ~default results =
+  match packages_that_were_not_found results with
+  | [] -> default
+  | [ pkg_name ] ->
+    User_error.raise
+      ([ Pp.textf
+           "%s was not found in the configured repositories."
+           (Dune_lang.Package_name.to_string pkg_name)
+       ]
+       @ default)
+  | pkg_names ->
+    User_error.raise
+      ([ Pp.text "The following packages were not found in the configured repositories:"
+       ; Pp.enumerate pkg_names ~f:(fun name ->
+           Dune_lang.Package_name.to_string name |> Pp.verbatim)
+       ]
+       @ default)
+;;
+
+let find ~repos ~local_packages { Lock_dir.packages; _ } =
```

Comment:
If this function only needs the packages, it would be better if it just took the packages as an argument.